master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala

```diff
@@ -678,16 +679,9 @@ private[celeborn] class Master(
         if (requestSlots.maxWorkers <= 0) slotsAssignMaxWorkers
         else Math.min(slotsAssignMaxWorkers, requestSlots.maxWorkers)),
       numAvailableWorkers)
-    val startIndex = Random.nextInt(numAvailableWorkers)
-    val selectedWorkers = new util.ArrayList[WorkerInfo](numWorkers)
-    selectedWorkers.addAll(availableWorkers.subList(
-      startIndex,
-      Math.min(numAvailableWorkers, startIndex + numWorkers)))
-    if (startIndex + numWorkers > numAvailableWorkers) {
-      selectedWorkers.addAll(availableWorkers.subList(
-        0,
-        startIndex + numWorkers - numAvailableWorkers))
-    }
+    Collections.rotate(availableWorkers, Random.nextInt(numAvailableWorkers))
+    val selectedWorkers = availableWorkers.subList(0, numWorkers)
+
     // offer slots
     val slots =
```

Comment:
With this change, where the first `numWorkers` in `availableWorkers` are chosen and rest are not, do we want to relook at how `offerSlotsExtraSize` is handled below ? And limit to the set outside of `[0, numWorkers)` ?

Reply:
>  And limit to the set outside of [0, numWorkers)

Hi, choose the `[0, numWorkers)` workers as `selectedWorkers`, it not means we will use all of these workers when offer slots.

Reply:
@mridulm I think there is no problem because the workers' snapshot is a new list.
```java
private def workersSnapShot: util.List[WorkerInfo] =
    statusSystem.workers.synchronized(new util.ArrayList[WorkerInfo](statusSystem.workers))
```
And the extra workers are selected from those workers who not participated in this shuffle.

```java
    val workersNotSelected = availableWorkers.asScala.filter(!slots.containsKey(_))
```